qurator/eynollah/processor.py

```diff
@@ -1,68 +1,61 @@
-                'curved_line': self.parameter['curved_line'],
-                'full_layout': self.parameter['full_layout'],
-                'allow_scaling': self.parameter['allow_scaling'],
-                'headers_off': self.parameter['headers_off'],
-                'tables': self.parameter['tables'],
-                'override_dpi': self.parameter['dpi'],
-                'logger': LOG,
-                'pcgts': pcgts,
-                'image_filename': image_filename
-                }
-            Eynollah(**eynollah_kwargs).run()
-            file_id = make_file_id(input_file, self.output_file_grp)
-            pcgts.set_pcGtsId(file_id)
-            self.workspace.add_file(
-                ID=file_id,
-                file_grp=self.output_file_grp,
-                pageId=page_id,
-                mimetype=MIMETYPE_PAGE,
-                local_filename=join(self.output_file_grp, file_id) + '.xml',
-                content=to_xml(pcgts))
+    def process_page_pcgts(self, *input_pcgts: Optional[OcrdPage], page_id: Optional[str] = None) -> OcrdPageResult:
+        assert input_pcgts
+        assert input_pcgts[0]
+        assert self.parameter
+        pcgts = input_pcgts[0]
+        page = pcgts.get_Page()
+        # if not('://' in page.imageFilename):
+        #     image_filename = next(self.workspace.mets.find_files(local_filename=page.imageFilename)).local_filename
+        # else:
+        #     # could be a URL with file:// or truly remote
+        #     image_filename = self.workspace.download_file(next(self.workspace.mets.find_files(url=page.imageFilename))).local_filename
+        # XXX loses DPI information
+        page_image, _, _ = self.workspace.image_from_page(
+            page, page_id,
+            # avoid any features that would change the coordinate system: cropped,deskewed
+            # (the PAGE builder merely adds regions, so afterwards we would not know which to transform)
+            # also avoid binarization as models usually fare better on grayscale/RGB
+            feature_filter='cropped,deskewed,binarized')
+        eynollah = Eynollah(
+            self.resolve_resource(self.parameter['models']),
+            self.logger,
+            allow_enhancement=self.parameter['allow_enhancement'],
+            curved_line=self.parameter['curved_line'],
+            light_version=self.parameter['light_mode'],
+            right2left=self.parameter['right_to_left'],
+            ignore_page_extraction=self.parameter['ignore_page_extraction'],
+            textline_light=self.parameter['textline_light'],
+            full_layout=self.parameter['full_layout'],
+            allow_scaling=self.parameter['allow_scaling'],
+            headers_off=self.parameter['headers_off'],
+            tables=self.parameter['tables'],
+            override_dpi=self.parameter['dpi'],
+            pcgts=pcgts,
+            image_filename=page.imageFilename,
+            image_pil=page_image
```

Comment:
Note: that filename might **not** be where that image came from in `workspace.image_from_page`. It could well be a derived image generated by some previous processor (just not a cropped, deskewed or binarized image, because that would have changed its coordinate system).

Reply:
It's still a bit hazy for me when `image_filename` is actually used. Ideally, `image_pil` should take preference and `image_filename` is only for the plotter/writer, at least in the "single image mode" we're using.

One of the aspects I hope I'll be able to improve a bit with https://github.com/qurator-spk/eynollah/tree/refactoring-2024-08/

Reply:
Perhaps we can also re-use `session` across Eynollah invokations in addition to `models`?

Reply:
In theory, yes, but with standalone eynollah being focused on batch processing now, I am honestly not sure how/where sessions are defined for the non-`dir_in` option - @vahidrezanezhad can you tell us?